hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/RefreshMetaProcedure.java

```diff
@@ -0,0 +1,416 @@
+      LOG.error("Failed to update the hbase:meta table", ioe);
+      throw ioe;
+    }
+  }
+
+  private Flow executeComplete() {
+    currentRegions = null;
+    latestRegions = null;
+    LOG.info("RefreshMetaProcedure completed successfully");
+    return Flow.NO_MORE_STATE;
+  }
+
+  /** Compares the current regions in hbase:meta with the latest regions from backing storage
+   * and applies necessary mutations (additions, deletions, or modifications) to the meta table.
+   *
+   * @param currentMap Current regions from hbase:meta
+   * @param latestMap Latest regions from backing storage
+   * @param connection HBase connection to use for meta table operations
+   * @param env MasterProcedureEnv for accessing master services
+   * @throws IOException If there is an error accessing the meta table or backing storage
+   */
+  private void compareAndUpdateRegions(Map<String, RegionInfo> currentMap, Map<String, RegionInfo> latestMap,
+    Connection connection, MasterProcedureEnv env) throws IOException {
+
+    List<Mutation> mutations = new ArrayList<>();
+
+    for (Map.Entry<String, RegionInfo> entry : latestMap.entrySet()) {
+      String regionId = entry.getKey();
+      RegionInfo latestRegion = entry.getValue();
+      RegionInfo currentRegion = currentMap.get(regionId);
+      if (currentRegion == null) {
+        mutations.add(MetaTableAccessor.makePutFromRegionInfo(latestRegion));
+        LOG.debug("Adding the region to meta: {}", latestRegion.getRegionNameAsString());
+      } else {
+        if (hasBoundaryChanged(currentRegion, latestRegion)) {
+          mutations.add(MetaTableAccessor.makePutFromRegionInfo(latestRegion));
+          LOG.debug("Adding a put to update region boundaries in meta: {}",
+            latestRegion.getRegionNameAsString());
+        }
+      }
+    }
+
+    for (Map.Entry<String, RegionInfo> entry : currentMap.entrySet()) {
+      String regionId = entry.getKey();
+      RegionInfo currentRegion = entry.getValue();
+      if (!latestMap.containsKey(regionId)) {
+        mutations.add(MetaTableAccessor.makeDeleteFromRegionInfo(currentRegion,
+          EnvironmentEdgeManager.currentTime()));
+        LOG.debug("Removing region from meta: {}", currentRegion.getRegionNameAsString());
+      }
+    }
+
+    if (!mutations.isEmpty()) {
+      LOG.info("Applying {} mutations to meta table", mutations.size());
+      executeBatchAndRetry(connection, mutations, env);
+    } else {
+      LOG.info("No update needed - meta table is in sync with backing storage");
+    }
+  }
+
+  private void executeBatchAndRetry(Connection connection, List<Mutation> mutations,
+    MasterProcedureEnv env) throws IOException {
+    List<List<Mutation>> chunks = Lists.partition(mutations, MUTATION_BATCH_SIZE);
+
+    for (int i = 0; i < chunks.size(); i++) {
+      List<Mutation> chunk = chunks.get(i);
+      RetryCounter retryCounter = ProcedureUtil.createRetryCounter(env.getMasterConfiguration());
```

Comment:
I see the following error handling in `InitMetaProcedure`:

```java
    } catch (IOException e) {
      if (retryCounter == null) {
        retryCounter = ProcedureUtil.createRetryCounter(env.getMasterConfiguration());
      }
      long backoff = retryCounter.getBackoffTimeAndIncrementAttempts();
      LOG.warn("Failed to init meta, suspend {}secs", backoff, e);
      setTimeout(Math.toIntExact(backoff));
      setState(ProcedureProtos.ProcedureState.WAITING_TIMEOUT);
      skipPersistence();
      throw new ProcedureSuspendedException();
    }
```

wdyt?
We could make the execute method a new step in the flow and let the procedure framework do the retry / backoff mechanism.